### PR TITLE
Plugins: Introduce HTTP 207 Multi Status response to api/ds/query

### LINKED ISSUE #45870: Instrumentation: Allow proper request instrumentation of /api/ds/query
Based on discussions in https://github.com/grafana/grafana/discussions/40444 we've identified a first iteration where we want to change how HTTP response statuses is returned from the /api/ds/query endpoint, used for querying data sources. To be able to properly monitor Grafana and measure SLI's we need to make sure that Grafana only returns non-2xx HTTP response status

**How it currently works:** 
- Returns status 500 Internal server error if querying a datasource returns an error https://github.com/grafana/grafana/blob/1cc5b25a0541abb04fa945cecd2b25ceac1586db/pkg/api/metrics.go#L25
- Returns status 200 OK if querying a datasource doesn't return an error and the response data by `refId` doesn't have the error property set https://github.com/grafana/grafana/blob/1cc5b25a0541abb04fa945cecd2b25ceac1586db/pkg/api/metrics.go#L91
- Returns status 400 Bad request if querying a datasource doesn't return an error and one of the response data by `refId` have the error property set  https://github.com/grafana/grafana/blob/1cc5b25a0541abb04fa945cecd2b25ceac1586db/pkg/api/metrics.go#L94

This makes it very hard to properly monitor and measure SLI's since a badly configured/behaving datasource adds negatively to the SLI's.

**Suggested change:**
Pre-query validation: Grafana backend validation of handling /api/ds/query request before calling backend plugin.

- If Grafana pre-query validation fails it could return 4xx - 5xx response status since it's an actual Grafana problem
- If Grafana pre-query validation succeeds and manage to connect to a backend plugin the response status should always be 207 Multi-Status

**Investigations:**
By changing so that response status 207 Multi-Status always is returned we observed the following:
- When using testdata datasource with 1 query (Randow walk) returns one frame with data it incorrectly shows a panel error "Query error: 207 Multi-Status" and renders the data.
- When using testdata datasource with 2 queries (Randow walk and Randow walk with error) it returns two responses, one with proper data frames and one with error it shows the warning/error in panel and renders the data as expected. 

The panel error "Query error: 207 Multi-Status" is probably shown due to code below and needs to be changed: https://github.com/grafana/grafana/blob/feae959c9da12885fc79786d57d7c415595c260a/packages/grafana-runtime/src/utils/queryResponse.ts#L126

In addition, the following datasources seems to depend on checking the HTTP response status and could potentially be other community ones so we need to handle this somehow:
- https://github.com/grafana/grafana/blob/1cc5b25a0541abb04fa945cecd2b25ceac1586db/public/app/plugins/datasource/cloudwatch/datasource.ts#L630
- https://github.com/grafana/grafana/blob/1cc5b25a0541abb04fa945cecd2b25ceac1586db/public/app/plugins/datasource/graphite/datasource.ts#L757

We need to consider suggested changes as a potential breaking change for anyone integrating with /api/ds/query:
- Grafana UI (see above)
- Grafana Machine Learning
- Custom integration, e.g. curl

With that in mind we probably want to introduce suggested changes behind a feature toggle and deprecate returning HTTP status 200/400 when a response is returned from querying a datasource.

**Related:**
The request metric middleware that adds HTTP request histogram/summaries probably should exclude the following endpoints (or handle them in a special way) because they're proxies and built to return the HTTP status from the downstream server:
 - `/api/datasources/proxy/:id/*`
 - `/api/datasources/:id/resources/*`
 - `/api/datasources/:id/health`
 - `/api/plugins/:pluginId/resources/*`
 - `/api/plugins/:pluginId/health`
 - `/api/plugin-proxy/*`
 - `/api/gnet/`

### PR DESCRIPTION
**Pull request by wbrowne**
**What this PR does / why we need it**:
Based on [this discussion](https://github.com/grafana/grafana/discussions/40444)

**Which issue(s) this PR fixes**:

Fixes #45870

**Special notes for your reviewer**:

### CODE REVIEW

File: conf/defaults.ini [hunk 1]
```diff
@@ -1149,6 +1149,9 @@ commandPalette = true
 # Use dynamic labels in CloudWatch datasource
 cloudWatchDynamicLabels = false
 
+# Use HTTP 207 Multi Status for data source query responses
+datasourceQueryMultiStatus = false
```
Comment (joshhunt) [inline conf/defaults.ini:1153]: I don't think default-false feature flags need to be defined in here, right? By default feature flags are already off.

Comment (wbrowne) [reply conf/defaults.ini:1153]: Ah good shout. Just blindly copy pasta 👀

File: docs/sources/http_api/data_source.md [hunk 2]
```diff
@@ -691,6 +691,16 @@ In addition, specific properties of each data source should be added in a reques
 }
 ```
 
+#### Status codes
+
+| Code | Description                                                                                                                                                            |
+| ---- | ---------------------------------------------------------------------------------------------------------------------------------------------------------------------- |
+| 200  | All data source queries returned a successful response.                                                                                                                |
+| 400  | Bad request. Invalid json, missing content-type, missing or invalid fields, etc. Or one or more data source queries were unsuccessful. Refer to body for more details. |
```
Comment (chri2547) [inline docs/sources/http_api/data_source.md:699]:
```suggestion
| 400  | Bad request due to invalid JSON, missing content type, missing or invalid fields, etc. Or one or more data source queries were unsuccessful. Refer to the body for more details. |
```

**Review by chri2547 — CHANGES_REQUESTED** [review]
One suggestion. Thank you!

**Review by marefr — APPROVED** [review]
LGTM

**Review by chri2547 — APPROVED** [review]
LGTM!

**Pull request merged**